juju/juju
# cmd/juju/storage: revise storage tabular format

## Description of change

Instead of having storage, volumes and filesystems,
we go back to having just "storage". It is not
helpful to separate storage from volume/filesystem;
instead, we add more information from the machine
storage to the unit storage entry.

New fields added to volume/filesystem/storage:
include "life" in the storage, volume, filesystem
details in YAML/JSON output. This can be useful for
debugging. Included the pool name in volume and
filesystem info. The pool name is useful for users
to see what class of storage is attached to a unit.

Also: fix a bug that would cause "juju storage" to
fail when a storage instance has no associated
volume or filesystem. This will be the case up
until the storage instance's attached unit is assigned
to a machine.

## QA steps

SCENARIO 1
1. juju bootstrap
2. juju deploy postgresql --storage pgdata=1G
3. juju storage --format=yaml
check that the storage includes life fields, status is colourised

The colour for status should be yellow for "detaching", "detached", and green for "attaching", "attached".

SCENARIO 2
same as scenario 1, but use juju 2.0 client in step 3
(check that it can handle the new fields.)

SCENARIO 3
same as scenario 1, but use juju 2.0 client in steps 1-2
(check that juju storage works, but the "Pool" column is omitted.)

## Documentation changes

The tabular format of "juju storage" has been updated. If there are any examples in the documentation, we should update them to match.

## Bug reference

None.
